rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp

```diff
@@ -0,0 +1,456 @@
+      }
+    });
+
+  // Get memory from SHM buffer if available.
+  if (shm_manager.has_value() &&
+    zc_shm_manager_check(&shm_manager.value()))
+  {
+    shmbuf = zc_shm_alloc(
+      &shm_manager.value(),
+      max_data_length);
+    if (!z_check(shmbuf.value())) {
+      zc_shm_gc(&shm_manager.value());
+      shmbuf = zc_shm_alloc(&shm_manager.value(), max_data_length);
+      if (!z_check(shmbuf.value())) {
+        // TODO(Yadunund): Should we revert to regular allocation and not return an error?
+        RMW_SET_ERROR_MSG("Failed to allocate a SHM buffer, even after GCing");
+        return RMW_RET_ERROR;
+      }
+    }
+    msg_bytes = reinterpret_cast<char *>(zc_shmbuf_ptr(&shmbuf.value()));
+  } else {
+    // Get memory from the allocator.
+    msg_bytes = static_cast<char *>(allocator->allocate(max_data_length, allocator->state));
+    RMW_CHECK_FOR_NULL_WITH_MSG(
+      msg_bytes, "bytes for message is null", return RMW_RET_BAD_ALLOC);
+  }
+
+  // Object that manages the raw buffer
+  eprosima::fastcdr::FastBuffer fastbuffer(msg_bytes, max_data_length);
+
+  // Object that serializes the data
+  rmw_zenoh_cpp::Cdr ser(fastbuffer);
+  if (!type_support_->serialize_ros_message(
+      ros_message,
+      ser.get_cdr(),
+      type_support_impl_))
+  {
+    RMW_SET_ERROR_MSG("could not serialize ROS message");
+    return RMW_RET_ERROR;
+  }
+
+  const size_t data_length = ser.get_serialized_data_length();
+
+  const int64_t sequence_number = sequence_number_++;
+
+  z_owned_bytes_map_t map =
+    create_map_and_set_sequence_num(sequence_number, gid_);
+  if (!z_check(map)) {
+    // create_map_and_set_sequence_num already set the error
+    return RMW_RET_ERROR;
+  }
+  auto always_free_attachment_map = rcpputils::make_scope_exit(
+    [&map]() {
+      z_bytes_map_drop(z_move(map));
+    });
+
+  int ret;
+  // The encoding is simply forwarded and is useful when key expressions in the
+  // session use different encoding formats. In our case, all key expressions
+  // will be encoded with CDR so it does not really matter.
+  z_publisher_put_options_t options = z_publisher_put_options_default();
+  options.attachment = z_bytes_map_as_attachment(&map);
+
+  if (shmbuf.has_value()) {
+    zc_shmbuf_set_length(&shmbuf.value(), data_length);
+    zc_owned_payload_t payload = zc_shmbuf_into_payload(z_move(shmbuf.value()));
+    ret = zc_publisher_put_owned(z_loan(pub_), z_move(payload), &options);
+  } else {
+    // Returns 0 if success.
+    ret = z_publisher_put(
+      z_loan(pub_),
+      reinterpret_cast<const uint8_t *>(msg_bytes),
+      data_length,
+      &options);
+  }
+  if (ret) {
+    RMW_SET_ERROR_MSG("unable to publish message");
+    return RMW_RET_ERROR;
+  }
+
+  return RMW_RET_OK;
+}
+
+///=============================================================================
+std::shared_ptr<const liveliness::Entity> PublisherData::entity() const
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+  return entity_;
```

Comment:
The `Entity` is read-only (no methods to change it's state) so it should be ok?

Reply:
The alternative would be to add APIs that expose methods of `Entity` and `EventsManager`. eg `PublisherData::guid()` which would return `entity_->guid()`.

Reply:
I've updated the implementation to not return`Entity::SharedPtr`. Instead, the required methods from `Entity` are exposed. This also improves the interaction with `GraphCache` since the `Entity` used to add/remove callbacks is technically a different one from the `Entity` created within `GraphCache` (from liveliness token). However, their `guids` are the same and the hashing was based on this `guid`. the `GraphCache` API now adds/removes callbacks with the `guids` which makes things cleaner.

I didn't do this with `EventsManager` for two reasons
- We need to cast `rmw_event->data` to a common type for both publisher and subscription datas. I can't think of a good way to do this given that `rmw_event->data` is `void *` type erased.
- The `EventManager` locks an [internal mutex](https://github.com/ros2/rmw_zenoh/blob/824886a4ea9fd8a032139fbb31d0181df1731407/rmw_zenoh_cpp/src/detail/event.hpp#L142) when accessing other data members so having an `EventsManager::SharedPtr` and using it across different threads should be safe.


Commit:
https://github.com/ros2/rmw_zenoh/pull/278/commits/071242996ee235ae5aed872274fed44dcf7f2808

Reply:
In general, I like what you've done with this change.  I have one more suggestion of something I think we should fix, then I think this will be good.  I'll make a separate comment about that.